### DISCUSSION TITLE
Problems migrating MDX support from Gatsby to Next.JS

### DISCUSSION THREAD

**Discussion by Protonull** [vercel/next.js#29720] [Help]
Hey, so I'm having a little trouble migrating from Gatsby to Next.JS in regards to mdx support, at least in terms of expected behaviour. Below is a segment of my Gatsby config:
```ts
// gatsby-config.ts
...
{
    resolve: "gatsby-plugin-mdx",
    options: {
        defaultLayouts: {
            default: require.resolve("./src/components/mdx-wrapper.tsx")
        },
    },
}
...
```
which references a particular component which wraps the raw content of the mdx file, and that's fantastic, it means I can define a `MDXProvider` and have very granular control over what and how markdown gets parsed and transformed into components.

Whereas, with Next.JS, I've followed the instructions [here](https://mdxjs.com/getting-started/next/) and it works, but I can't find a way to specify a wrapper component so it's rendering with standard HTML tags. I've read [here](https://mdxjs.com/guides/wrapper-customization/) and [here](https://github.com/vercel/next.js/discussions/21780) that you can just wrap everything in your `_app.js` with a `MDXProvider` but that seems more like a hack than a genuine solution and seems to pre-suppose that all your pages and posts will be MDX files. I'm trying to find if there's a solution in the configuration but that's pretty opaque and haven't found any documentation on it, at least insofar as a comparable solution to `gatsby-plugin-mdx`'s default layouts.

**Comment by saso124** [comment]
what is the benefit of Next.js?

**Comment by ifonefox** [comment]
In the App component, you can use the [`useRouter`](https://nextjs.org/docs/api-reference/next/router) hook to see if you need to wrap the page or not. Another method would to use [Nextjs's dynamic import](https://nextjs.org/docs/advanced-features/dynamic-import). Here is an [example](https://github.com/uu64/nextjs-mdx-blog-sample/blob/master/pages/blog/%5B...id%5D.tsx) I found.

Instead of using Nextjs's mdx loader, you could also use [`next-mdx-remote`](https://github.com/hashicorp/next-mdx-remote), a popular community-made library that sends the compiled MDX file using `getStaticProps`/`getServerSideProps`, and gives you more flexibility on the page layout and `MDXProvider`. There is an [article on the official Nextjs blog](https://nextjs.org/blog/markdown) that shows how to use it.

**Comment by Protonull** [comment]
> what is the benefit of Next.js?

While I agree that this comment is not on topic, its giving me flashbacks to:
\> How do I do <thing> in JavaScript?
\> Just use jQuery

Gatsby is very good don't get me wrong but it's also very presumptive. Something I quite like about next.js is that you can use `getStaticProps` *within the same file* to retrieve data, and you can use whatever retrieval method you want, whereas with Gatsby, you're strong-armed into using GraphQL or otherwise be relegated to `gatsby-node.js` where you have to wrestle with endless callbacks. Perhaps I'm old school, but fs'ing a folder is a lot easier than fudging around with GraphQL queries and sorting through their responses.

Switching to next.js has meant that my project has less than half the number of dependencies; has a shallower folder structure; you can retrieve data how you want; that static-rendering is native and exceptionally easy; etc. I haven't even been using next.js for very long.
